bridge/jira/export.go

```diff
@@ -0,0 +1,455 @@
+				// so we use the comment ID plus the timestamp of the update, as
+				// reported by JIRA. Note that this must be consistent with the importer
+				// during ensureComment()
+				id = fmt.Sprintf("%s-%d", comment.ID, comment.Updated.Unix())
+			}
+
+		case *bug.SetStatusOperation:
+			jiraStatus, hasStatus := statusMap[opr.Status.String()]
+			if hasStatus {
+				exportTime, err = UpdateIssueStatus(client, bugJiraID, jiraStatus)
+				if err != nil {
+					err := errors.Wrap(err, "editing status")
+					out <- core.NewExportError(err, b.Id())
+					// Failure to update status isn't necessarily a big error. It's
+					// possible that we just don't have enough information to make that
+					// update. In this case, just don't export the operation.
+					continue
+				}
+				out <- core.NewExportStatusChange(op.Id())
+				id = bugJiraID
+			} else {
+				out <- core.NewExportError(fmt.Errorf(
+					"No jira status mapped for %.8s", opr.Status.String()), b.Id())
+			}
+
+		case *bug.SetTitleOperation:
+			exportTime, err = client.UpdateIssueTitle(bugJiraID, opr.Title)
+			if err != nil {
+				err := errors.Wrap(err, "editing title")
+				out <- core.NewExportError(err, b.Id())
+				return err
+			}
+			out <- core.NewExportTitleEdition(op.Id())
+			id = bugJiraID
+
+		case *bug.LabelChangeOperation:
+			exportTime, err = client.UpdateLabels(
+				bugJiraID, opr.Added, opr.Removed)
+			if err != nil {
+				err := errors.Wrap(err, "updating labels")
+				out <- core.NewExportError(err, b.Id())
+				return err
+			}
+			out <- core.NewExportLabelChange(op.Id())
+			id = bugJiraID
+
+		default:
+			panic("unhandled operation type case")
+		}
+
+		// mark operation as exported
+		err = markOperationAsExported(
+			b, op.Id(), id, self.project.Key, exportTime)
+		if err != nil {
+			err := errors.Wrap(err, "marking operation as exported")
+			out <- core.NewExportError(err, b.Id())
+			return err
+		}
+
+		// commit at each operation export to avoid exporting same events multiple
+		// times
+		err = b.CommitAsNeeded()
+		if err != nil {
+			err := errors.Wrap(err, "bug commit")
+			out <- core.NewExportError(err, b.Id())
+			return err
+		}
+	}
+
+	return nil
+}
+
+func markOperationAsExported(
+	b *cache.BugCache, target entity.Id, jiraID, jiraProject string,
+	exportTime time.Time) error {
+
+	newMetadata := map[string]string{
+		keyJiraID:      jiraID,
+		keyJiraProject: jiraProject,
+	}
+	if !exportTime.IsZero() {
+		newMetadata[keyJiraExportTime] = exportTime.Format(http.TimeFormat)
+	}
+
+	_, err := b.SetMetadata(target, newMetadata)
+	return err
+}
+
+// UpdateIssueStatus attempts to change the "status" field by finding a
+// transition which achieves the desired state and then performing that
+// transition
+func UpdateIssueStatus(
+	client *Client, issueKeyOrID string, desiredStateNameOrID string) (
+	time.Time, error) {
+
+	var responseTime time.Time
+
+	tlist, err := client.GetTransitions(issueKeyOrID)
```

Comment:
really not an emergency but this could be cached at some point

Reply:
Indeed (bug `8acce9c`), but it's not *exactly* simple. What we can do is download the full state graph ("workflow") but then we'll need to include logic to figure out what transitions are actually currently valid for an issue. OR we can just ask JIRA which transitions are currently valid (that's what this API endpoint does). 

Not going round-trip through the REST API would be faster, but it is also *slightly* racy in that the cached workflow *might* be stale and it would require git-bug implementation of the validation logic that JIRA implements. In practice workflows don't change frequently so it's probably the better choice, but there is some actual design work to be done there so I'm holding off on this.

Reply:
yeah, let see if this is actually a performance problem first.